header/headertest/verify_test.go

```diff
@@ -15,52 +16,64 @@ func TestVerify(t *testing.T) {
 	trusted, untrustedAdj, untrustedNonAdj := h[0], h[1], h[2]
 	tests := []struct {
 		prepare func() *header.ExtendedHeader
-		err     bool
+		err     error
 	}{
 		{
 			prepare: func() *header.ExtendedHeader { return untrustedAdj },
-			err:     false,
+			err:     nil,
 		},
 		{
 			prepare: func() *header.ExtendedHeader {
 				return untrustedNonAdj
 			},
-			err: false,
+			err: nil,
 		},
 		{
 			prepare: func() *header.ExtendedHeader {
 				untrusted := *untrustedAdj
 				untrusted.ValidatorsHash = tmrand.Bytes(32)
 				return &untrusted
 			},
-			err: true,
+			err: &libhead.VerifyError{
+				Reason: header.ErrValidatorHashMismatch,
+			},
 		},
 		{
 			prepare: func() *header.ExtendedHeader {
 				untrusted := *untrustedAdj
 				untrusted.RawHeader.LastBlockID.Hash = tmrand.Bytes(32)
 				return &untrusted
 			},
-			err: true,
+			err: &libhead.VerifyError{
+				Reason: header.ErrLastHeaderHashMismatch,
+			},
 		},
 		{
 			prepare: func() *header.ExtendedHeader {
 				untrusted := *untrustedNonAdj
 				untrusted.Commit = NewTestSuite(t, 2).Commit(RandRawHeader(t))
 				return &untrusted
 			},
-			err: true,
+			err: &libhead.VerifyError{
+				Reason: header.ErrVerifyCommitLightTrustingFailed,
+			},
 		},
 	}
 
 	for i, test := range tests {
 		t.Run(strconv.Itoa(i), func(t *testing.T) {
 			err := trusted.Verify(test.prepare())
-			if test.err {
-				assert.Error(t, err)
-			} else {
+			if test.err == nil {
 				assert.NoError(t, err)
+				return
+			}
+			if err == nil {
+				t.Errorf("expected err: %v, got nil", test.err)
+				return
 			}
+			reason := err.(*libhead.VerifyError).Reason
+			testReason := test.err.(*libhead.VerifyError).Reason
+			assert.ErrorIs(t, reason, testReason)
```

Comment:
```suggestion
			assert.ErrorIs(t, errors.Unwrap(test.err), err)
```

This all can be simplified to

Reply:
Even for success case it should check that nil == nil

Reply:
Ah, we can actually remove the `errors.Unwrap` if we don't wrap the expected test case errors with `VerifyError`. We test for inner error first of all, we don't care much if its wrapped or not

Reply:
The first test checks for the success case where nil == nil. This is the code that enters ```If `test.err` == nil```  and then asserts that `err` is not an error which basically checks if it's nil. 

I tried adding your suggestion but it breaks the test both with `errors.Unwrap` and without it.

Reply:
> I tried adding your suggestion but it breaks the test both with errors.Unwrap and without it so added errors.Unwrap to both instead which works fine with nil arguments as well.

For this to work you need to apply the following suggestion
> if we don't wrap the expected test case errors with VerifyError

Instead of this
```
err: &libhead.VerifyError{
	Reason: header.ErrVerifyCommitLightTrustingFailed,
}
```

This
```
err: header.ErrVerifyCommitLightTrustingFailed,
```